download definitions while generating to speed up plan and update

grosser opened the pull request:
before (latest release):
```
time bundle exec rake kennel:plan PROJECT=foo
Generating ... 1.51s
Storing ... 0.01s
Downloading definitions ... 7.09s
Diffing ... 0.3s
Plan:
Nothing to do

real	0m10.006s
user	0m3.593s
sys	0m0.551s
```

after:
```
time bundle exec rake kennel:plan PROJECT=foo
Downloading definitions ...
Finding parts ...
Finding parts ... 1.56s
Building json ... 0.01s
Downloading definitions ... 6.6s
Diffing ... 0.1s
Plan:
Nothing to do

real	0m7.228s
user	0m3.455s
sys	0m0.552s
```

rework of https://github.com/grosser/kennel/pull/216

@zdrve

zdrve commented:
Will take a look once the tests pass.

zdrve commented:
Doesn't give any advantage if you `rake generate plan` though, does it?

zdrve commented on lib/kennel.rb:
```diff
@@ -80,12 +80,26 @@ def update
     private
 
+    def download_definitions
+      Progress.progress("Downloading definitions", plain: true) do
```
The PR description shows this line appearing first and list, sandwiching the `generate` stuff. However that's not guaranteed; there's a race condition. Combine that with the fact that this progress is "plain" but not all of the inner ones are, and the output might end up looking odd.

Maybe that's fine though.

grosser replied:
- there should be no inner progress since there are only these 2 in parallel
- they will appear in a random order, not really a big issue afaik

grosser commented:
why would anyone `generate plan` when plan already generates ?

zdrve commented:
I don't know what to make of this PR yet. The description is misleading, and the tests don't pass. Ping me again when it's ready.

zdrve commented:
It's also unclear to me why you felt the need to rework my PR. Yes I know mine isn't ready, but you seem to have taken a whole different approach, I'd love to hear the rationale.

grosser commented:
- I made a fresh PR because I though there might be a simpler way of achieving the same goal
- how is the description misleading ?
- tests don't pass in https://github.com/grosser/kennel/pull/216 either and passing tests should not be needed to evaluate the complexity/merit of a PRs approach, afaik this is ready and just needs test fixup unless you can point out any flaws in the approach in which case I might rework or scrap it
- this PR moves the download code and runs it in parallel, to me this seems much simpler than "extract new class, manage threads to know the state of that class, check on the class in multiple places"

zdrve commented:
re. Misleading: which commit did the "before" text come from?

re: 216 not passing either: yes but I'm not _claiming_ it's ready. It's still draft.

grosser commented:
before="before this change" -> master

grosser commented:
fixed up the tests

grosser commented on lib/kennel/tasks.rb:
```diff
@@ -102,7 +102,7 @@ def git_push?
   # also generate parts so users see and commit updated generated automatically
   desc "show planned datadog changes (scope with PROJECT=name)"
-  task plan: :generate do
+  task plan: :environment do
```
was a bug before, called `generate` twice, did not matter since it was cached but makes the parallelism break now

zdrve replied:
So when we merge this we need to point out to people (and add to the change notes) that `rake plan` used to write to `generated`, but now doesn't.

grosser replied:
Everything will still work.
`rake plan` writes to `generated` as before.
`rake generate plan` was always redundant.
It might be good to add a warning/error to anyone running `plan` and `generate` in a single command since that would be slower than just running `plan`, but only "as slow as before".

zdrve replied:
With the PR as it stands at the moment, `rake generate plan` will (I believe) run generate, plan, and then generate again, which presumably has worse performance.

Not necessarily the end of the world; the answer might be "well don't do that then".

Or, we _consciously_ change what `plan` does, so that it _doesn't_ generate (like it used to).

(Fundamentally the problem here is that using Rake tasks is always going to make this kind of optimisation painful).

zdrve commented:
> before="before this change" -> master

I do not believe that the "before" log example is accurate.

zdrve commented:
Because of the impact it has on the end user (change of behaviour of "plan", and the need to run tasks in the right order), I'm not yet convinced that this approach is simpler. But I get that it's subjective.

grosser commented:
There is no change to `rake plan`.
No ordering is required.
`before` is not from the "merge-base", but "latest release", but the time is the same.

zdrve commented:
> before is not from the "merge-base"

OK, can you please fix the PR description then?

zdrve commented:
> There is no change to rake plan.

Yes there is. `rake plan` used to update `./generated`; now it doesn't.

grosser commented:
- aaah I mixed up `generate` and `generated` 🤦
- updated `before` description
- added `generate` back into `plan` so everything should behave as before

idk how widespread `generate plan` usage is (afaik it's not documented anywhere), but we can add a warning or error for that to discourage it if

grosser commented:
@zdrve all good ?

zdrve commented on test/kennel/syncer_test.rb:
```diff
@@ -609,7 +614,7 @@ def expect_gets(answer)
       it "refuses to update tracking on resources with ids since they would be deleted by other updates" do
         expected << monitor("a", "b", foo: "bar", id: 123)
-        monitors << monitor_api_response("a", "b", id: 123).merge(message: "An innocent monitor -- Managed by kennel b:b")
+        monitors << monitor_api_response("a", "b", id: 123).merge(message: "An innocent monitor -- Managed by kennel b:b", tracking_id: "b:b")
```
Isn't this the same as: ?

```suggestion
        monitors << monitor_api_response("b", "b", id: 123)
```

grosser replied:
yeah, good catch ... idk how I ended in that rabbithole
